diff --git a/control/controlhttp/client.go b/control/controlhttp/client.go
index 06a2131fd..e9fe30b96 100644
--- a/control/controlhttp/client.go
+++ b/control/controlhttp/client.go
@@ -36,6 +36,7 @@
 	"sync/atomic"
 	"time"
 
+	"github.com/coder/websocket"
 	"tailscale.com/control/controlbase"
 	"tailscale.com/control/controlhttp/controlhttpcommon"
 	"tailscale.com/envknob"
@@ -48,6 +49,7 @@
 	"tailscale.com/net/netx"
 	"tailscale.com/net/sockstats"
 	"tailscale.com/net/tlsdial"
+	"tailscale.com/net/wsconn"
 	"tailscale.com/syncs"
 	"tailscale.com/tailcfg"
 	"tailscale.com/tstime"
@@ -94,6 +96,7 @@ func (a *Dialer) httpsFallbackDelay() time.Duration {
 }
 
 var _ = envknob.RegisterBool("TS_USE_CONTROL_DIAL_PLAN") // to record at init time whether it's in use
+var forceWebSocket = func() bool {return envknob.BoolDefaultTrue("TS_FORCE_CONTROL_WEBSOCKET")} // use WebSocket by default; set to "0" to fallback to HTTP/HTTPS
 
 func (a *Dialer) dial(ctx context.Context) (*ClientConn, error) {
 
@@ -237,6 +240,9 @@ func (a *Dialer) dialHost(ctx context.Context) (*ClientConn, error) {
 // If optAddr is valid, then no DNS is used and the connection will be made to the
 // provided address.
 func (a *Dialer) dialHostOpt(ctx context.Context, optAddr netip.Addr, optACEHost string) (*ClientConn, error) {
+	if forceWebSocket() {
+		return a.dialWebSocket(ctx, optAddr, optACEHost)
+	}
 	// Create one shared context used by both port 80 and port 443 dials.
 	// If port 80 is still in flight when 443 returns, this deferred cancel
 	// will stop the port 80 dial.
@@ -576,3 +582,44 @@ func (a *Dialer) tryURLUpgrade(ctx context.Context, u *url.URL, optAddr netip.Ad
 
 	return netutil.NewAltReadWriteCloserConn(rwc, switchedConn), nil
 }
+
+func (a *Dialer) dialWebSocket(ctx context.Context, optAddr netip.Addr, optACEHost string) (*ClientConn, error) {
+	init, cont, err := controlbase.ClientDeferred(a.MachineKey, a.ControlKey, a.ProtocolVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	wsScheme := "wss"
+	host := a.Hostname
+	if a.HTTPPort != "" && a.HTTPPort != "80" && a.HTTPSPort == "443" {
+		wsScheme = "ws"
+		host = net.JoinHostPort(host, a.HTTPPort)
+	} else if a.HTTPSPort != "" && a.HTTPSPort != "443" {
+		host = net.JoinHostPort(host, a.HTTPSPort)
+	}
+
+	wsURL := &url.URL{
+		Scheme: wsScheme,
+		Host:   host,
+		Path:   serverUpgradePath,
+		RawQuery: url.Values{
+			controlhttpcommon.HandshakeHeaderName: []string{base64.StdEncoding.EncodeToString(init)},
+		}.Encode(),
+	}
+
+	wsConn, _, err := websocket.Dial(ctx, wsURL.String(), &websocket.DialOptions{
+		Subprotocols: []string{controlhttpcommon.UpgradeHeaderValue},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	netConn := wsconn.NetConn(context.Background(), wsConn, websocket.MessageBinary, wsURL.String())
+
+	cbConn, err := cont(ctx, netConn)
+	if err != nil {
+		netConn.Close()
+		return nil, err
+	}
+	return &ClientConn{Conn: cbConn}, nil
+}
\ No newline at end of file
